### ISSUE TITLE
[java] StringInstantiation: False-positive when using `new String(charArray)`

### ISSUE THREAD

**Issue by boris-petrov** [pmd/pmd#4874]
**Affects PMD Version:**

7.0.0

**Rule:**

[StringInstantiation](https://pmd.github.io/pmd/pmd_rules_java_performance.html#stringinstantiation)

**Description:**

```java
var charArray = new char[] { 'a', 'b' };
var string = new String(charArray);
```

This gives a warning while it shouldn't.

**Code Sample demonstrating the issue:**

See above.

**Expected outcome:**

No violation.

**Running PMD through:**

Gradle 8.7, JDK 21.0.2, Linux.

**Comment by boris-petrov** [comment]
@adangel, @jsotuyod I still get that error in PMD 7.2.0. The code is a bit different:

```java
return switch (foo) {
	case char[] array -> new String(array);
	case String string -> string;
	default -> throw new Exception();
};
```

**Comment by oowekyala** [comment]
This appears to be a problem with variable scoping (related to pattern matching) rather than a problem with this specific rule... I'll look into it and create a ticket
